hail/src/main/scala/is/hail/io/RowStore.scala

```diff
@@ -986,343 +935,224 @@ object EmitPackDecoder {
-          mainCode.append(s"if (!${call}) { s = ${r1}; goto pull; }\n")
-        case t: PFloat64 =>
-          val call = if (skip) "this->skip_double()" else s"this->decode_double((double*)${addr})"
-          mainCode.append(s"if (!${call}) { s = ${r1}; goto pull; }\n")
-
-        case t: PBinary =>
-          // TBinary - usually a string - has an int length, followed by that number of bytes
-          val ptr = stateVar("ptr", depth)
-          val len = stateVar("len", depth)
-          val idx = stateVar("idx", depth)
-          mainCode.append(s"if (!this->decode_length(&${len})) { s = ${r1}; goto pull; }\n")
-          if (skip) {
-            mainCode.append(s"for (${idx} = 0; ${idx} < ${len};) {\n")
-            val r2 = allocState(s"${name}.bytes");
-            mainCode.append(s"  auto ngot = this->skip_bytes(${len}-${idx});\n")
-            mainCode.append(s"  if (ngot <= 0) { s = ${r2}; goto pull; }\n")
-            mainCode.append(s"  ${idx} += ngot;\n")
-            mainCode.append(s"}\n")
-          } else {
-            mainCode.append(s"${ptr} = region->allocate(4, 4+${len});\n")
-            mainCode.append(s"*(char**)${addr} = ${ptr};\n")
-            mainCode.append(s"*(int32_t*)${ptr} = ${len};\n")
-            mainCode.append(s"for (${idx} = 0; ${idx} < ${len};) {\n")
-            val r2 = allocState(s"${name}.bytes");
-            mainCode.append(s"  auto ngot = this->decode_bytes(${ptr}+4+${idx}, ${len}-${idx});\n")
-            mainCode.append(s"  if (ngot <= 0) { s = ${r2}; goto pull; }\n")
-            mainCode.append(s"  ${idx} += ngot;\n")
-            mainCode.append(s"}\n")
-          }
+  def decodeBinary(input_buf_ptr: cxx.Expression, region: cxx.Expression, off: cxx.Expression): cxx.Code = {
+    val len = cxx.Variable("len", "int", s"$input_buf_ptr->read_int()")
+    val boff = cxx.Variable("boff", "char *", s"$region->allocate(${PBinary.contentAlignment}, $len + 4)")
+    s"""${len.define}
+       |${boff.define}
+       |store_address($off, $boff);
+       |store_int($boff, $len);
+       |$input_buf_ptr->read_bytes($boff + 4, $len);""".stripMargin
+  }
+
+  def decodeArray(t: PArray, rt: PArray, input_buf_ptr: cxx.Expression, region: cxx.Expression, off: cxx.Expression): cxx.Code = {
+    val len = cxx.Variable("len", "int", s"$input_buf_ptr->read_int()")
+    val sab = new cxx.StagedContainerBuilder(region, rt)
+    var decodeElt = decode(t.elementType, rt.elementType, input_buf_ptr, region, cxx.Expression(sab.eltOffset))
+    if (rt.elementType.required)
+      decodeElt =
+        s"""if (!${rt.cxxIsElementMissing(sab.end(), sab.idx)}) {
+           |  $decodeElt
+           |}
+         """.stripMargin
+
+    s"""${ len.define }
+       |${ sab.start(len, clearMissing = false) }
+       |store_address($off, ${sab.end()});
+       |${ if (rt.elementType.required) "" else s"$input_buf_ptr->read_bytes(${sab.end()} + 4, ${rt.cxxNMissingBytes(s"$len")});" }
```

Comment:
Can we put parens around the if body?  I worry that if we modify the while to be multiple statements, this will break.  We should probably do this consistently like we do with parens for associatively.